### PR TITLE
Basis block height cache

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: lib/LedgerNodeBlocks.js [hunk 1]
```diff
@@ -1,18 +1,37 @@
...
+ * Copyright (c) 2017-2019 Digital Bazaar, Inc. All rights reserved.
  */
 'use strict';
 
-const {callbackify} = require('bedrock').util;
+const bedrock = require('bedrock');
+const {config, util: {callbackify}} = bedrock;
+const LRU = require('lru-cache');
+
+let lruCache;
```
Comment (davidlehn) [inline lib/LedgerNodeBlocks.js]: Maybe a more descriptive name like `blockHeightCache` would be better?  Short file, so not that big a deal.

Comment (mattcollier) [reply lib/LedgerNodeBlocks.js]: Done.

File: lib/LedgerNodeBlocks.js [hunk 2]
```diff
@@ -51,6 +70,29 @@ module.exports = class LedgerNodeBlocks {
...
+   *
+   * @return {Promise<Number>} the latest block height for the ledger.
+   */
+  async getLatestBlockHeight({useCache = false} = {}) {
+    const {ledgerNodeId} = this;
+    let basisBlockHeight;
+    if(useCache) {
+      basisBlockHeight = blockHeightCache.get(ledgerNodeId);
+    }
+    if(basisBlockHeight !== undefined) {
```
Comment (davidlehn) [inline lib/LedgerNodeBlocks.js]: This basisBlockHeight check block could be moved up inside the useCache check block.

Comment (mattcollier) [reply lib/LedgerNodeBlocks.js]: Done.

**Comment by dlongley** [comment]
This patch seems like it could become problematic in the future. If we ever have an instance that can service significantly more than 500 ledger nodes that see regular traffic, this cache may be getting wiped out constantly. That seems like a very possible future. It also means that each instance is listening for events from every ledger node (I'm not sure if that's the first time we've done this or not).

So I'm not sure if this is a good idea. If we do feel like we need to proceed with it then I think like we need some kind of documentation about some impending problems with the approach.

**Comment by dlongley** [comment]
Should we put this in the redis cache and update that cache whenever we write blocks (vs. having an event listener) instead?

**Comment by mattcollier** [comment]
This is primarily intended to improve the hotpath for setting `basisBlockHeight` on new operations.  Currently there is no cache and a more expensive getLatestSummary API call is being used rather than the covered query being used here.

Write up your apocalyptic vision in an issue perhaps?

**Comment by mattcollier** [comment]
> Should we put this in the redis cache and update that cache whenever we write blocks (vs. having an event listener) instead?

It's already in the cache, and I believe you recommended an LRU cache yourself which I think IS a good idea.  It certainly works well in the common case.

**Comment by mattcollier** [comment]
Oh another reason redis is not involved here is because there is no requirement to have redis at all in the ledger core.  All redis requirements are in Continuity at present.

**Comment by mattcollier** [comment]
Is the `block.add` bit a requirement at all?  What if that bit is removed, is that better?  The TTL for the cache is supposed to ensure freshness.

**Comment by mattcollier** [comment]
replaced with https://github.com/digitalbazaar/bedrock-ledger-node/pull/70

**Pull request closed without merging**